File: javascript/mergeSort.js
//an example of merge sort algorithm
function mergeSort(arr1, arr2) {
    let newArr = [];
    let i = 0;
    let j = 0;
    //loop through arrays until one is exhausted of elements
    while (i < arr1.length && j < arr2.length) {
        if (arr1[i] <= arr2[j]) {
            newArr.push(arr1[i]);
            i++;
        } else {
            newArr.push(arr2[j]);
            j++
        }
    }
    //loop through the elements of the remaining array
    if (i === arr1.length) {
        while (j !== arr2.length) {
            newArr.push(arr2[j]);
            j++;
        }
    } else {
        while (i !== arr1.length) {
            newArr.push(arr1[i])
            i++
        }
    }
    return newArr
}

function divideAndConquer(array) {
    //slice array elements until they have 1 element by dividing array in half
    //resulting in newArray - 1 half and newArrayTwo - the other half
    if (array.length > 1) {
        var index = Math.floor(array.length / 2);
        var newArray = array.slice(0, index);
        var newArrayTwo = array.slice(index, array.length);
        if (newArray.length > 1) {
            newArray = divideAndConquer(newArray);
        }
        if (newArrayTwo.length > 1) {
            newArrayTwo = divideAndConquer(newArrayTwo);
        }
    }
    return mergeSort(newArray, newArrayTwo);
}

var result = divideAndConquer([1, 2, 32, 4, 8, 9, 110, 6, 6, 7, 9, 8, 13]);
console.log(result);

File: C++/merge_sort.cpp
#include<bits/stdc++.h>
using namespace std;

int merge(int arr[],int l,int r,int mid)
{
    int l1=0,l2=0,i=0,j=0,k=0,arr1[100],arr2[100];
    l1=mid-l+1;
    l2=r-mid;
    for(i=l,k=0;i<=mid;i++,k++)
    arr1[k]=arr[i];
    for(i=mid+1,k=0;i<=r;i++,k++)
    arr2[k]=arr[i];
    i=0,j=0,k=l;
    while(i<l1&&j<l2)
    {
        if(arr1[i]<=arr2[j])
        {
            arr[k]=arr1[i];
            i++;
        }
        else
        {
            arr[k]=arr2[j];
            j++;
        }
        k++;
    }
    for(;i<l1;i++,k++)
    arr[k]=arr1[i];
    for(;j<l2;j++,k++)
    arr[k]=arr2[j];
}

int merge_sort(int arr[],int l,int r)
{
    if(l<r)
    {
        int mid=l+(r-l)/2;
        merge_sort(arr,l,mid);
        merge_sort(arr,mid+1,r);
        merge(arr,l,r,mid);
    }
}

int main()
{
    int  arr[]={4,6,3,2,9,7,1,6,5},i=0,n=0;
    //n=sizeof(arr)/sizeof(arr[0]);
    merge_sort(arr,0,9-1);
    for(i=0;i<9;i++)
    cout<<arr[i]<<" ";
    return 0;
}